azurelinuxagent/common/cgroupapi.py

```diff
@@ -455,52 +480,57 @@ def start_extension_command(self, extension_name, command, shell, cwd, env, stdo
             env=env,
             preexec_fn=os.setsid)
 
-        # Wait a bit and check if we completed with error
-        time.sleep(1)
-        return_code = process.poll()
-
-        if return_code is not None and return_code != 0:
-            process_output = read_output(stdout, stderr)
-
-            # When systemd-run successfully invokes a command, thereby creating its unit, it will output the
-            # unit's name. Since the scope name is only known to systemd-run, and not to the extension itself,
-            # if scope_name appears in the output, we are certain systemd-run managed to run.
-            if scope_name not in process_output:
-                add_event(AGENT_NAME,
-                          version=CURRENT_VERSION,
-                          op=WALAEventOperation.InvokeCommandUsingSystemd,
-                          is_success=False,
-                          message='Failed to run systemd-run for unit {0}.scope. '
-                                  'Process exited with code {1} and output {2}'.format(scope_name,
-                                                                                       return_code,
-                                                                                       process_output))
-                # Reset the stdout and stderr
-                stdout.truncate(0)
-                stderr.truncate(0)
-
-                # Try starting the process without systemd-run
-                process = subprocess.Popen(
-                    command,
-                    shell=shell,
-                    cwd=cwd,
-                    env=env,
-                    stdout=stdout,
-                    stderr=stderr,
-                    preexec_fn=os.setsid)
-
-                return process, []
-
-        cgroups = []
-
         logger.info("Started extension using scope '{0}'", scope_name)
+        extension_cgroups = []
 
         def create_cgroup(controller):
             cgroup_path = os.path.join(CGROUPS_FILE_SYSTEM_ROOT, controller, 'system.slice', scope_name + ".scope")
-            cgroups.append(CGroup.create(cgroup_path, controller, extension_name))
-
-        self._foreach_controller(create_cgroup, 'Cannot create cgroup for extension {0}; resource usage will not be tracked.'.format(extension_name))
-
-        return process, cgroups
+            extension_cgroups.append(CGroup.create(cgroup_path, controller, extension_name))
+
+        self._foreach_controller(create_cgroup, 'Cannot create cgroup for extension {0}; '
+                                                'resource usage will not be tracked.'.format(extension_name))
+        self.track_cgroups(extension_cgroups)
+
+        # Wait for process completion or timeout
+        timed_out, return_code = wait_for_process_completion_or_timeout(process, timeout)
```

Comment:
can you check the case when the extension completes with failure (i.e. timed_out == False, return_code != 0)

also, the series of if statements below are a little hard to follow? maybe we can re-arrange them somehow to make very obvious that all conditions are covered?

Thanks

Reply:
Any extension-related failures (timing out or returning != 0) are handled by the `start_subprocess_and_wait_for_completion` method in `processutil`. This is method handles a "normal" extension invocation (without systemd) and is used in several places: in the FileSystemCGroupAPI, in `exthandlers.py` when cgroups are disabled, and in the fallback option when systemd doesn't work.

Reply:
I stand corrected! The case when the extension fails when invoked by systemd was not covered. Sending an update now. Thanks for catching it.